Automattic/jetpack
# Subscriptions: Add user content link tracking and redirection for links in emails.

Fixes https://github.com/Automattic/wp-calypso/issues/80864

## Proposed changes:

When user clicks on links from post content in emails, this link now will point to a page hosted on the site's own hostname.

In that page, a hidden iframe will be rendered, pointing to `subscribe.wordpress.com`, where tracking (for email link click stats) is performed, and the final destination URL is then sent from this iframe child to the parent window (via postMessage).

The parent window will then redirect the user to the final redirection URL.

### Other information:


## Jetpack product discussion
pdDOJh-2i1-p2

## Does this pull request change what data or activity we track or use?
It will allow us to track links in the body of the posts and comments, that we couldn't track until now. More info on this here: pdDOJh-1X2-p2

## Testing instructions:

**Setup your site:**
1. Setup Jurassic Tube.
2. Ensure your Jetpack site is connected to WordPress.com.
3. Ensure Subscriptions feature is enabled (Jetpack > Settings > Newsletter).

**Enable blog sticker:**
1. Go to Blog RC for this jurassic tube site.
2. Enable `user-content-link-tracking` blog sticker.

**Send yourself an email:**
1. Publish a post containing links in the post content.
2. Check the new post email received in your inbox.
3. The links in the post content in the email should be pointing to your own domain, e.g. `your_site_here.jurassic.tube/?action=user_content_redirect&all=other&params=here`.

**Test the redirection link:**
1. Click on the link (or copy & paste) the link to your browser.
2. It should redirect you to the final destination url.
3. Go to `/stats/email/clicks/day/<post_id>/your_site_here.jurassic.tube` on Calypso.
4. You should see the stats of the user content links at the bottom.

**Test with subscriptions module disabled:**
1. Disable subscriptions module via (Jetpack > Settings > Newsletter).
2. Enter the url again.
3. You'll land on your site's main page.

![image](https://github.com/Automattic/jetpack/assets/1287077/64d73c96-c70e-47a2-92cd-ba27938214b6)
